src/vec.rs

```diff
@@ -876,60 +807,730 @@ impl<T, const N: usize> Vec<T, N> {
+    ///
+    /// let mut v: Vec<u8, 16> = Vec::new();
+    /// v.extend_from_slice(&[1, 2, 3]).unwrap();
+    /// ```
+    #[inline]
+    #[allow(clippy::result_unit_err)]
+    pub fn from_slice(other: &[T]) -> Result<Self, ()>
+    where
+        T: Clone,
+    {
+        let mut v = Vec::new();
+        v.extend_from_slice(other)?;
+        Ok(v)
     }
-}
 
-impl<const N: usize> fmt::Write for Vec<u8, N> {
-    fn write_str(&mut self, s: &str) -> fmt::Result {
-        match self.extend_from_slice(s.as_bytes()) {
-            Ok(()) => Ok(()),
-            Err(_) => Err(fmt::Error),
+    /// Constructs a new vector with a fixed capacity of `N`, initializing
+    /// it with the provided array.
+    ///
+    /// The length of the provided array, `M` may be equal to _or_ less than
+    /// the capacity of the vector, `N`.
+    ///
+    /// If the length of the provided array is greater than the capacity of the
+    /// vector a compile-time error will be produced.
+    pub fn from_array<const M: usize>(src: [T; M]) -> Self {
+        // Const assert M >= 0
+        crate::sealed::greater_than_eq_0::<M>();
+        // Const assert N >= M
+        crate::sealed::greater_than_eq::<N, M>();
+
+        // We've got to copy `src`, but we're functionally moving it. Don't run
+        // any Drop code for T.
+        let src = ManuallyDrop::new(src);
+
+        if N == M {
+            Self {
+                len: N,
+                // NOTE(unsafe) ManuallyDrop<[T; M]> and [MaybeUninit<T>; N]
+                // have the same layout when N == M.
+                buffer: unsafe { mem::transmute_copy(&src) },
+            }
+        } else {
+            let mut v = Vec::<T, N>::new();
+
+            for (src_elem, dst_elem) in src.iter().zip(v.buffer.iter_mut()) {
+                // NOTE(unsafe) src element is not going to drop as src itself
+                // is wrapped in a ManuallyDrop.
+                dst_elem.write(unsafe { ptr::read(src_elem) });
+            }
+
+            v.len = M;
+            v
+        }
+    }
+
+    /// Clones a vec into a new vec
+    pub(crate) fn clone(&self) -> Self
+    where
+        T: Clone,
+    {
+        let mut new = Self::new();
+        // avoid `extend_from_slice` as that introduces a runtime check/panicking branch
+        for elem in self {
+            unsafe {
+                new.push_unchecked(elem.clone());
+            }
+        }
+        new
+    }
+
+    /// Get a reference to the Vec, erasing the `N` const-generic
+    ///
+    /// This can also be used through type coerction, since `Vec<T, N>` implements `Unsize<VecView<T>>`:
+    ///
+    /// ```rust
+    /// use heapless::{Vec, VecView};
+    ///
+    /// let vec: Vec<u8, 10> = Vec::from_slice(&[1, 2, 3, 4]).unwrap();
+    /// let view: &VecView<_> = &vec;
+    /// ```
+    pub const fn as_view(&self) -> &VecView<T> {
+        self
+    }
+
+    /// Get a `mut` reference to the Vec, erasing the `N` const-generic
```

Comment:
```suggestion
    /// Get a mutable reference to the `Vec`, erasing the `N` const-generic.
```

Reply:
I'm not sure I like replacing `mut` with "mutable", since `mut` doesn't necessarily mean "mutable", but rather means "exclusive", and a non-mut reference is shared.

Not a hill I'm ready to die on though.

Reply:
> since `mut` doesn't necessarily mean "mutable"

`mut` is an abbreviation for mutable, so it does mean that quite literally, in addition to meaning “exclusive” in the context of Rust references.

Reply:
Here is an example of `mut` being called an "exclusive" : https://google.github.io/comprehensive-rust/references/exclusive.html

Exclusive is more in line with the existence of interior mutability, but exclusive doesn't necessarily map to the `mut` keyword, because it's a shorthand for "mutable", so I prefer referring to the keyword directly, so there is no ambiguity at all, and everyone still understands it.